cpu/esp32/esp-now/esp_now_netif.c

```diff
@@ -53,97 +53,151 @@ static int _send(gnrc_netif_t *netif, gnrc_pktsnip_t *pkt)
 
     if (netif_hdr->flags & (GNRC_NETIF_HDR_FLAGS_BROADCAST | GNRC_NETIF_HDR_FLAGS_MULTICAST)) {
         /* ESP-NOW does not support multicast, always broadcast */
-        memset(esp_now_pkt.mac, 0xff, ESP_NOW_ADDR_LEN);
+        memset(mac, 0xff, ESP_NOW_ADDR_LEN);
     } else if (netif_hdr->dst_l2addr_len == ESP_NOW_ADDR_LEN) {
-        memcpy(esp_now_pkt.mac, gnrc_netif_hdr_get_dst_addr(netif_hdr), ESP_NOW_ADDR_LEN);
+        memcpy(mac, gnrc_netif_hdr_get_dst_addr(netif_hdr), ESP_NOW_ADDR_LEN);
     } else {
         DEBUG("gnrc_esp_now: destination address had unexpected format"
               "(flags=%d, dst_l2addr_len=%d)\n", netif_hdr->flags, netif_hdr->dst_l2addr_len);
+        gnrc_pktbuf_release(pkt);
         return -EBADMSG;
     }
 
+    iolist_t payload_iolist = {
+        .iol_base = calloc(1, sizeof(esp_now_pkt_hdr_t)),
+        .iol_len = sizeof(esp_now_pkt_hdr_t)
+    };
+
+    iolist_t iolist = {
+        .iol_base = mac,
+        .iol_len = sizeof(mac),
+        .iol_next = &payload_iolist
+    };
+
+    esp_now_pkt_hdr_t *esp_hdr = payload_iolist.iol_base;
+
+    if (!esp_hdr) {
+        DEBUG("gnrc_esp_now: no memory to allocate payload buffer\n");
+        gnrc_pktbuf_release(pkt);
+        return -ENOMEM;
+    }
+
     switch (payload->type) {
 #ifdef MODULE_GNRC_SIXLOWPAN
         case GNRC_NETTYPE_SIXLOWPAN:
-            esp_now_pkt.buf.hdr.flags = 1;
+            esp_hdr->flags = 1;
             break;
 #endif
         default:
-            esp_now_pkt.buf.hdr.flags = 0;
+            esp_hdr->flags = 0;
     }
 
-    iolist_t iolist = {
-        .iol_base = (char*)&esp_now_pkt,
-        .iol_len = sizeof(esp_now_pkt_t)
-    };
-
-    unsigned payload_len = 0;
-    uint8_t *pos = esp_now_pkt.buf.data;
-
     while (payload) {
-        payload_len += payload->size;
-
-        if (payload_len > ESP_NOW_MAX_SIZE) {
+        if (payload_iolist.iol_len + payload->size > ESP_NOW_MAX_SIZE) {
             DEBUG("gnrc_esp_now: payload length exceeds maximum(%u>%u)\n",
-                  payload_len, ESP_NOW_MAX_SIZE);
+                  payload_iolist.iol_len, ESP_NOW_MAX_SIZE);
+            free(payload_iolist.iol_base);
             gnrc_pktbuf_release(pkt);
             return -EBADMSG;
         }
 
-        memcpy(pos, payload->data, payload->size);
-        pos += payload->size;
+        payload_iolist.iol_base = realloc(payload_iolist.iol_base,
```

Comment:
Is it really necessary to work with `realloc` and `free`? Your first version worked without dynamic memory allocation. According to [Best Practice for RIOT Programming](https://github.com/RIOT-OS/RIOT/wiki/Best-Practice-for-RIOT-Programming) static memory allocation should be preferred.

Reply:
The alternative would be either a 256 byte sized static buffer or a 256 byte buffer on the stack.
Since pretty much all the existing netif drivers are using (re)alloc already(indirectly via gnrc_pktbuf_* class of functions), I decided for this variant. Could also use a pktbuf here instead, which would under the hood do the same thing though.

The underlying problem is that esp_now_send needs to get a single buffer with the full packet. And the netif needs to prepend the packet header to it, so the only way to do that is to copy the whole thing to another buffer.

Reply:
Thanks for the explanation.